### PR TITLE
Add new route implementation

### PR DESCRIPTION
**Pull request by alexander-schranz**
| Q | A
| --- | ---
| Bug fix? | no
| New feature? |yes
| BC breaks? | yes
| Deprecations? | yes 
| Fixed tickets | fixes # 
| Related issues/PRs | see https://github.com/sulu/sulu/issues/7672
| License | MIT
| Documentation PR | sulu/sulu-docs# 

#### What's in this PR?

Add new route implementation.

#### Why?

I did some experimenting to simplify the route handling in https://github.com/alexander-schranz/route-prototype. Instead of that the developer need manually care about updating things we will work with doctrine listeners which will update child urls automatically to avoid issues like:

 - https://github.com/sulu/sulu/issues/7580
 - https://github.com/sulu/sulu/issues/4990

Also we will not longer have a difference between unpublished or published URLs. Currently in Article Bundle I could create 2 articles with the url `/test` and `/test` and it did only error when try to publish one. This should not longer be the case as we already reserving and create a unique URL when saving the Entity and does not end in publishing errors specially when try using in combination with the AutomationBundle.

#### ToDo

 - [x] Create Entity
 - [x] Childs are update
 - [x] Grand childs are update
 - [x] Multi Localization support
 - [x] Multi Site support
 - [x] Support for non Site connected routes (article)
    - [x] Site A with `/test` and Site B `/test` and article parent Site A `/test/article` should not be updated when Site B `/test` was changed

#### ToDos in own Pull Requests

 - [x] History Handling https://github.com/sulu/sulu/pull/7830
 - [x] Add Multi Site tests https://github.com/sulu/sulu/pull/7831
 - [x] Add Multi Localization https://github.com/sulu/sulu/pull/7831
 - [x] Add CrossSite Site test (article (site: null) -> page (site: website)) https://github.com/sulu/sulu/pull/7834
 - [x] Add Test for url with same prefix are not effect on the same level https://github.com/sulu/sulu/pull/7836
 - [x] Tree Full Edit: https://github.com/sulu/sulu/pull/7835
 - [x] Route for AutoIncrement Entities: https://github.com/sulu/sulu/pull/7859
 - [x] Dynamic Router https://github.com/sulu/sulu/pull/7860
 - [ ] Locale Switcher https://github.com/sulu/sulu/pull/7874
 - [ ] Implement in Content Bundle https://github.com/sulu/sulu/pull/7749

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -36,7 +36,7 @@
         "doctrine/cache": "^1.12.1 || ^2.0",
         "doctrine/collections": "^1.6 || ^2.0",
         "doctrine/data-fixtures": "^1.4",
-        "doctrine/dbal": "^3.3",
+        "doctrine/dbal": "^3.6",
```
Comment (alexander-schranz) [inline composer.json:39]: > Deprecated `Connection::PARAM_*_ARRAY` constants: Use the corresponding constants on `ArrayParameterType` instead.

So I also replaced it in AccessControlRepository

The 3.6 was released 2 years ago: https://github.com/doctrine/dbal/releases/3.6.0 so we are fine to upgrade.

File: src/Sulu/Bundle/SecurityBundle/Entity/AccessControlRepository.php [hunk 2]
```diff
@@ -87,7 +88,7 @@ public function findIdsWithGrantedPermissions(
             ->andWhere('accessControl.role IN (' . $systemRoleQueryBuilder->getDQL() . ')')
             ->setParameter('system', $system)
             ->setParameter('entityClass', $entityClass)
-            ->setParameter('entityIds', $entityIds, Connection::PARAM_STR_ARRAY)
+            ->setParameter('entityIds', $entityIds, ArrayParameterType::STRING)
```
Comment (alexander-schranz) [inline src/Sulu/Bundle/SecurityBundle/Entity/AccessControlRepository.php:90]: DoctrineDBAL:

> Deprecated `Connection::PARAM_*_ARRAY` constants: Use the corresponding constants on `ArrayParameterType` instead.

File: packages/route/config/doctrine/Route/Route.orm.xml [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+            <unique-constraint name="ro_routes_resource_unique" fields="site,locale,resourceKey,resourceId" />
+        </unique-constraints>
+
+        <id name="id" type="integer" column="id">
+            <generator strategy="AUTO"/>
+        </id>
+
+        <field name="site" type="string" length="16" nullable="true"/>
+        <field name="locale" type="string" length="11"/>
+        <field name="slug" type="string"/>
```
Comment (alexander-schranz) [inline packages/route/config/doctrine/Route/Route.orm.xml]: To support or versions of mysql site + locale + slug can not be longer together as `191` chars. Which means in this case 191 - 16 -11 = 164 chars for the slug.

There would be one workaround to support longer versions of URL by using a additional `slugHash` field based on widely supported `SHA1` algorithm. That way we can support longer Urls 164.

@benr77 already requested that the current `title` of 64 is to small for his usecases. If we increase that to 128 or 191. A longer parent title and a child page or 3 levels of navigation could already hit the 191 limit and we would require trim the url correctly. But that is more a theoretical issue. Which may already existed before.

Comment (alexander-schranz) [reply packages/route/config/doctrine/Route/Route.orm.xml]: Interesting the workaround with a `Hash` is also documented in MySQL docs itself but as a alternative for composite indexes so n our case it would include the site, locale, slug: https://dev.mysql.com/doc/refman/8.4/en/multiple-column-indexes.html

File: packages/route/src/Infrastructure/Doctrine/EventListener/RouteChangedUpdater.php [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+
+            // update child and grand routes
+            $updateQueryBuilder = $connection->createQueryBuilder()
+                ->update($routesTableName, 'r')
+                ->set('slug', 'CONCAT(:newSlug' . $newSlugCast . ', SUBSTRING(slug, ' . (\strlen($oldSlug) + 1) . '))')
+                ->setParameter('newSlug', $newSlug, ParameterType::STRING)
+                ->where('parent_id IN (:parentIds)')
+                ->setParameter('parentIds', $parentIds, ArrayParameterType::INTEGER);
+
+            $updateQueryBuilder->executeStatement();
```
Comment (alexander-schranz) [inline packages/route/src/Infrastructure/Doctrine/EventListener/RouteChangedUpdater.php]: Not ideal handling of Postgres but works now as expected and is covers by tests.

Else the update query ends with follorwing error:

> Doctrine\DBAL\Exception\DriverException: An exception occurred while executing a query: SQLSTATE[42P18]: Indeterminate datatype: 7 ERROR:  could not determine data type of parameter $1

File: packages/route/config/doctrine/Route/Route.orm.xml [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+            <unique-constraint name="ro_routes_unique" fields="site,locale,slug" />
+            <unique-constraint name="ro_routes_resource_unique" fields="site,locale,resourceKey,resourceId" />
+        </unique-constraints>
+
+        <id name="id" type="integer" column="id">
+            <generator strategy="AUTO"/>
+        </id>
+
+        <field name="site" type="string" length="16" nullable="true"/>
+        <field name="locale" type="string" length="11"/>
```
Comment (Prokyonn) [inline packages/route/config/doctrine/Route/Route.orm.xml]: I guess we should increase that also to `15`

File: packages/route/src/Domain/Model/Route.php [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+    private ?string $locale;
+
+    private ?string $slug;
+
+    private ?Route $parentRoute;
+
+    private ?string $resourceKey;
+
+    private ?string $resourceId;
```
Comment (Prokyonn) [inline packages/route/src/Domain/Model/Route.php]: Why are these all nullable?

File: packages/route/config/doctrine/Route/Route.orm.xml (hunk 5 above)
Comment (alexander-schranz) [reply packages/route/config/doctrine/Route/Route.orm.xml]: changed the column length have to adopted the site and slug length to match into a index of 191. added a test case so we not accidently increase it.

File: packages/route/src/Domain/Model/Route.php (hunk 6 above)
Comment (alexander-schranz) [reply packages/route/src/Domain/Model/Route.php]: good catch, they should not.

File: packages/route/src/Domain/Model/Route.php [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+    {
+        $this->slug = $slug;
+
+        return $this;
+    }
+
+    public function setParentRoute(?Route $parentRoute): void
+    {
+        $this->parentRoute = $parentRoute;
+    }
```
Comment (Prokyonn) [inline packages/route/src/Domain/Model/Route.php]:
```suggestion
    public function setParentRoute(?Route $parentRoute): static
    {
        $this->parentRoute = $parentRoute;
        
        return $this;
    }
```

File: packages/route/src/Infrastructure/Doctrine/EventListener/RouteChangedUpdater.php [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+use Doctrine\DBAL\ParameterType;
+use Doctrine\DBAL\Platforms\PostgreSQLPlatform;
+use Doctrine\ORM\Event\OnClearEventArgs;
+use Doctrine\ORM\Event\PostFlushEventArgs;
+use Doctrine\ORM\Event\PreUpdateEventArgs;
+use Sulu\Route\Domain\Model\Route;
+use Symfony\Contracts\Service\ResetInterface;
+
+/**
+ * @internal No BC promise are given for this class. Can be changed or removed at any time.
```
Comment (Prokyonn) [inline packages/route/src/Infrastructure/Doctrine/EventListener/RouteChangedUpdater.php]:
```suggestion
 * @internal No BC promises are given for this class. It may be changed or removed at any time.
```

File: packages/route/src/Infrastructure/Doctrine/Repository/RouteRepository.php [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+
+use Doctrine\ORM\EntityManagerInterface;
+use Doctrine\ORM\EntityRepository;
+use Doctrine\ORM\Query;
+use Doctrine\ORM\QueryBuilder;
+use Sulu\Route\Domain\Model\Route;
+use Sulu\Route\Domain\Repository\RouteRepositoryInterface;
+
+/**
+ * @phpstan-import-type RouteFilter from \Sulu\Route\Domain\Repository\RouteRepositoryInterface
```
Comment (Prokyonn) [inline packages/route/src/Infrastructure/Doctrine/Repository/RouteRepository.php]:
```suggestion
 * @phpstan-import-type RouteFilter from RouteRepositoryInterface
```

I think you can omit the full path

File: packages/route/src/Domain/Model/Route.php [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+ *
+ * (c) Sulu GmbH
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Sulu\Route\Domain\Model;
+
+class Route
```
Comment (Prokyonn) [inline packages/route/src/Domain/Model/Route.php:17]: Does it make sense to implement an interface to make it better extendable? 🤔

Comment (alexander-schranz) [reply packages/route/src/Domain/Model/Route.php:17]: Would currently not yet support that. adding later interface isn't a bc break here. currently finalize the class.

Comment (alexander-schranz) [inline packages/route/src/Domain/Model/Route.php:17]:
```suggestion
final class Route
```

File: packages/route/src/Domain/Model/Route.php [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+    {
+        $this->slug = $slug;
+
+        return $this;
+    }
+
+    public function setParentRoute(?Route $parentRoute): static
+    {
+        $this->parentRoute = $parentRoute;
+
```
Comment (alexander-schranz) [inline packages/route/src/Domain/Model/Route.php]:
```suggestion

```

File: packages/route/src/Domain/Model/Route.php [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+    {
+        $this->slug = $slug;
+
+        return $this;
+    }
+
+    public function setParentRoute(?Route $parentRoute): static
+    {
+        $this->parentRoute = $parentRoute;
+        return $this;
```
Comment (alexander-schranz) [inline packages/route/src/Domain/Model/Route.php]:
```suggestion
        $this->parentRoute = $parentRoute;

        return $this;
```

File: packages/route/src/Domain/Model/Route.php [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+ *
+ * (c) Sulu GmbH
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Sulu\Route\Domain\Model;
+
+final class Route
```
Comment (alexander-schranz) [inline packages/route/src/Domain/Model/Route.php]:
```suggestion
/**
 * @final
 */
class Route
```

**Review by Prokyonn — APPROVED** [review]

**Comment by alexander-schranz** [comment]
8.4 error is unrelated to the PR. Looks like some issues in one of our dependencies maybe only a dev anomaly but we should test what is causing it symfony or the 2fa bundle.
